src/main/scala/io/picnicml/doddlemodel/cluster/DBSCAN.scala

```diff
@@ -0,0 +1,83 @@
+package io.picnicml.doddlemodel.cluster
+
+import breeze.linalg.functions.euclideanDistance
+import cats.syntax.option._
+import io.picnicml.doddlemodel.data.Features
+import io.picnicml.doddlemodel.typeclasses.Clusterer
+
+/** An immutable DBSCAN model.
+  *
+  * @param eps: the maximum distance between points in a group
+  * @param minSamples: the minimum number of point in a core group
+  *
+  * Examples:
+  * val model = DBSCAN()
+  * val model = DBSCAN(eps = 1.5)
+  * val model = DBSCAN(minSamples = 3)
+  * val model = DBSCAN(eps = 2.0, minSamples = 3)
+  */
+case class DBSCAN private(eps: Double, minSamples: Int, private val label: Option[Array[Int]])
+
+object DBSCAN {
+
+  def apply(eps: Double = 1.0, minSamples: Int = 1): DBSCAN = {
+    require(eps > 0.0, "Maximum distance needs to be larger than 0")
+    require(minSamples > 0, "Minimum number of samples needs to be larger than 0")
+    DBSCAN(eps, minSamples, none)
+  }
+
+  implicit lazy val ev: Clusterer[DBSCAN] = new Clusterer[DBSCAN] {
+
+    override def isFitted(model: DBSCAN): Boolean = model.label.isDefined
+
+    override protected def labelSafe(model: DBSCAN): Array[Int] = model.label.get
+
+    override protected def copy(model: DBSCAN): DBSCAN =
+      model.copy()
+
+    override protected def copy(model: DBSCAN, label: Array[Int]): DBSCAN =
+      model.copy(label = label.some)
+
+    override protected def fitSafe(model: DBSCAN, x: Features): DBSCAN = {
+      val label = Array.fill[Int](x.rows)(Int.MaxValue)
+      var groupId = -1
+      for (pointId <- 0 until x.rows if label(pointId) == Int.MaxValue) {
+        var groupQueue = findNeighbors(pointId, x, model.eps)
+        if (groupQueue.size + 1 < model.minSamples) {
+          label(pointId) = -1
```

Comment:
In my opinion, the code would be nicer if you got rid of the "magic numbers". 
As far as I can tell, `-1` represents noise and `Int.MaxValue`represents an unassigned point, correct?

I think defining two constants (for example `NOISE` and `UNASSIGNED`) and using them in code could improve readability. The constant for noise should be publicly accessible (so that users can do something like `result(i) == DBSCAN.NOISE`), while the other one can stay private (no point can be unassigned after fitting is done)